[Ready] Batched cholesky decomposition

vishwakftw opened the pull request:
Implements batching for the Cholesky decomposition.

Performance could be improved with a dedicated batched `tril` and `triu` op, which is also impeding autograd operations.

Changes made:
- batching code
- tests in `test_torch.py`, `test_cuda.py` and `test_autograd.py`.
- doc string modification
- autograd modification
- removal of `_batch_potrf` in `MultivariateNormal`.

vishwakftw committed:
Batched cholesky decomposition

ezyang commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -284,4 +301,67 @@ Tensor& potrs_out(Tensor& result, const Tensor& self, const Tensor& A, bool uppe
+static void apply_cholesky(Tensor& self, bool upper, std::vector<int64_t>& infos) {
+#ifndef USE_LAPACK
+  AT_ERROR("cholesky: LAPACK library not found in compilation");
+#endif
```
nit: `#else` out the rest of the code?

ezyang commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -284,4 +301,67 @@ Tensor& potrs_out(Tensor& result, const Tensor& self, const Tensor& A, bool uppe
+    return at::empty_like(self);
+  }
+  if (self.dim() == 2) {
+    return at::_th_potrf_single(self, upper);
```
So, at the end of the day, `_th_potrf_single` is calling the same LAPACK function, right? Can we just get rid of that codepath entirely and kill all of the TH bindings too? :)

ezyang replied:
(Totes ok if that happens in another PR.)

vishwakftw replied:
Yep, I think this should be possible. I'll get working on this soon.

ezyang commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -284,4 +301,67 @@ Tensor& potrs_out(Tensor& result, const Tensor& self, const Tensor& A, bool uppe
+  squareCheckInputs(self);
+
+  // TODO: Once triu, tril is implemented for batched tensors, this
+  // can be simplified.
```
nit: Maybe just write down explicitly what the simplification here is.

vishwakftw replied:
Soon, will do.

ezyang reviewed (approved):
Looks legit, thank you!

vishwakftw committed:
CR

- #else out sections that will not be reached
- Add detailed comment about using batched triu / tril and the resulting simplification

vishwakftw commented:
@ezyang I have addressed comments and resolved merge conflicts too.

vishwakftw commented:
@pytorchbot retest this please

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batch_cholesky

vishwakftw commented:
Build failures are unrelated.

bddppq commented on test/test_torch.py:
```diff
@@ -5334,6 +5334,29 @@ def test_cholesky(self):
+            A = cast(random_symmetric_pd_matrix(n, *batch_dims))
+            cholesky_exp = torch.stack([m.cholesky(upper=upper) for m in A.reshape(-1, n, n)])
+            cholesky_exp = cholesky_exp.reshape_as(A)
+            print(torch.cholesky(A, upper=upper))
```
Should these debug lines be removed?

vishwakftw replied:
Oh, I missed them. I am sorry about this. I’ll send in a patch.
